ext/bindings.go

```diff
@@ -94,3 +157,74 @@ func celBind(mef cel.MacroExprFactory, target ast.Expr, args []ast.Expr) (ast.Ex
 		resultExpr,
 	), nil
 }
+
+func newBlockScope(slotExprs []interpreter.Interpretable, expr interpreter.Interpretable) *blockScope {
+	bs := &blockScope{
+		slotExprs: slotExprs,
+		expr:      expr,
+	}
+	bs.slotActivationPool = &sync.Pool{
+		New: func() any {
+			sa := &slotActivation{
+				slotExprs: slotExprs,
+				slotVals:  make([]ref.Val, len(slotExprs)),
+			}
+			return sa
+		},
+	}
+	return bs
+}
+
+type blockScope struct {
+	slotExprs          []interpreter.Interpretable
+	expr               interpreter.Interpretable
+	slotActivationPool *sync.Pool
+}
+
+func (bs *blockScope) ID() int64 {
+	return bs.expr.ID()
+}
+
+func (bs *blockScope) Eval(activation interpreter.Activation) ref.Val {
+	sa := bs.slotActivationPool.Get().(*slotActivation)
+	sa.Activation = activation
+	defer bs.clearSlots(sa)
+	return bs.expr.Eval(sa)
+}
+
+func (bs *blockScope) clearSlots(sa *slotActivation) {
+	sa.reset()
+	bs.slotActivationPool.Put(sa)
+}
+
+type slotActivation struct {
+	interpreter.Activation
+	slotExprs []interpreter.Interpretable
+	slotVals  []ref.Val
+}
+
+func (sa *slotActivation) ResolveName(name string) (any, bool) {
```

Comment:
As discussed offline, we should add a recursive check here to ensure the sub expression we are evaluating is not the slot being resolved, to prevent bad expressions from infinite looping and killing the stack.

Reply:
Done, PTAL